### PR TITLE
[SYCL][clang] Emit default template arguments in integration header

### PR DESCRIPTION
**Pull request by Fznamznon**
For free function kernels support clang forward declares the kernel itself as well as its parameter types. In case a free function kernel has a parameter that is templated and has a default template argument, all template arguments including arguments that match default arguments must be printed in kernel's forward declarations, for example

```
template <typename T, typename = int> struct Arg {
  T val;
};

// For the kernel
SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(
    (ext::oneapi::experimental::nd_range_kernel<1>)) void foo(Arg<int> arg) {
  arg.val = 42;
}

// Integration header must contain
void foo(Arg<int, int> arg);
```

Unfortunately, even though integration header emission already has extensive support for forward declarations priting, some modifications to clang's type printing are still required, since neither of existing PrintingPolicy flags help to reach the correct result.
Using `SuppressDefaultTemplateArgs = true` doesn't help without printing canonical types, printing canonical types for the case like
```
template <typename T>
SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(
    (ext::oneapi::experimental::nd_range_kernel<1>)) void foo(Arg<T> arg) {
  arg.val = 42;
}
// Printing canonical types is causing the following integration header
template <typename T>
void foo(Arg<type-parameter-0-0, int> arg);
```

Using `SkipCanonicalizationOfTemplateTypeParms` field of printing policy doesn't help here since at the one point where it is checked we take canonical type of `Arg`, not its parameters and it will contain template argument types in canonical type after that.

### CODE REVIEW

**Review by lbushi25 — APPROVED** [review]
The tests for work group memory arguments used with free function kernels are now passing with these changes so for me this looks good.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 1]
```diff
@@ -6471,16 +6471,32 @@ void SYCLIntegrationHeader::emit(raw_ostream &O) {
...
-      ParmList += Param->getType().getCanonicalType().getAsString();
+      ParmList += Param->getType().getCanonicalType().getAsString(Policy);
     }
     FunctionTemplateDecl *FTD = K.SyclKernel->getPrimaryTemplate();
     Policy.SuppressDefinition = true;
     Policy.PolishForDeclaration = true;
+    Policy.FullyQualifiedName = true;
+    Policy.EnforceScopeForElaboratedTypes = true;
+
+    // SuppressDefaultTemplateArguments is a downstream addition that suppresses
```
Comment (elizabethandrews) [inline clang/lib/Sema/SemaSYCL.cpp:6488]: I think an example in the comments would be helpful here. Probably the same one in PR description. Just reading these is very confusing without context.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 2]
```diff
@@ -6471,16 +6471,32 @@ void SYCLIntegrationHeader::emit(raw_ostream &O) {
...
+    // true to emit function declaration that won't cause any compilation errors
+    // when present in the integration header.
+    // SuppressDefaultTemplateArgs is a flag from community code which adds
+    // suppression of template arguments that match default template arguments
+    // in template-ids printing.
+    Policy.SuppressDefaultTemplateArguments = true;
+    // EnforceDefaultTemplateArgs is a downstream addition that forces printing
+    // template arguments that match default template arguments while printing
+    // template-ids, even if the source code doesn't referece them.
+    Policy.EnforceDefaultTemplateArgs = true;
```
Comment (elizabethandrews) [inline clang/lib/Sema/SemaSYCL.cpp:6499]: The names make this quite confusing. 
`Policy.SuppressDefaultTemplateArguments` and `Policy.EnforceDefaultTemplateArgs` sounds like they do opposite things but are both set to `true` here. I assume `EnforceDefaultTemplateArgs` is technically the opposite of community `SuppressDefaultTemplateArgs` ? What does `match default template arguments` mean?

Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp:6499]: > I assume EnforceDefaultTemplateArgs is technically the opposite of community SuppressDefaultTemplateArgs ?

Technically, yes.

> What does match default template arguments mean?

A template argument in a template-id and with the same value as default argument in the template declaration that this template-id refers to.

Policy.SuppressDefaultTemplateArguments suppresses default template arguments from definition of template, i.e.
if function decl has
```
template <typename = int> void Imafreefunction(....) {...}
```
`typename = int` will be suppressed leaving simply `typename` in the integration header. Perhaps I should rename this flag to something like `Policy.SuppressDefaultTemplateArgumentsInDeclarations`? WDYT? It is a flag added by intel/llvm anyway.

File: clang/lib/Sema/SemaSYCL.cpp (hunk 1 above)
Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp:6501]: Added, thanks!

File: clang/lib/Sema/SemaSYCL.cpp [hunk 3]
```diff
@@ -6471,16 +6471,32 @@ void SYCLIntegrationHeader::emit(raw_ostream &O) {
...
+    // default template arguments in function declarations. It should be set to
+    // true to emit function declaration that won't cause any compilation errors
+    // when present in the integration header.
+    // SuppressDefaultTemplateArgs is a flag from community code which adds
+    // suppression of template arguments that match default template arguments
+    // in template-ids printing.
+    Policy.SuppressDefaultTemplateArguments = true;
+    // EnforceDefaultTemplateArgs is a downstream addition that forces printing
+    // template arguments that match default template arguments while printing
+    // template-ids, even if the source code doesn't referece them.
```
[the remark this answers is not in the document]
Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp]: Fixed, thanks!

**Review by premanandrao — APPROVED** [review]

**Comment by Fznamznon** [comment]
@elizabethandrews , are you ok with the change?

File: clang/lib/Sema/SemaSYCL.cpp (hunk 2 above)
Comment (elizabethandrews) [reply clang/lib/Sema/SemaSYCL.cpp:6525]: Thanks for explaining! I think you can leave it as-is

**Review by elizabethandrews — APPROVED** [review]

**Comment by Fznamznon** [comment]
@intel/llvm-gatekeepers , the PR is ready for submit. Please merge.

**Pull request merged**